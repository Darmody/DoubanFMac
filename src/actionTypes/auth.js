export const VERIFY_REQUEST = 'AUTH/VERIFY_REQUEST';
export const VERIFY_SUCCESS = 'AUTH/VERIFY_SUCCESS';
export const VERIFY_FAILURE = 'AUTH/VERIFY_FAILURE';
export const LOGIN_REQUEST = 'AUTH/LOGIN_REQUEST';
export const LOGIN_SUCCESS = 'AUTH/LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'AUTH/LOGIN_FAIL';
export const LOGOUT = 'AUTH/LOGOUT';
