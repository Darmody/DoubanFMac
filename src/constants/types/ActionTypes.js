// auth
export const LOGIN_REQUEST: ActionType = 'LOGIN/REQUEST'
export const LOGIN_SUCCESS: ActionType = 'LOGIN/SUCCESS'
export const LOGIN_FAILURE: ActionType = 'LOGIN/FAILURE'

// users
export const USER_CURRENT_REQUEST: ActionType = 'USER/CURRENT/REQUEST'

// entities
export const ENTITIES_SAVE: ActionType = 'ENTITIES/SAVE'
export const ENTITIES_UPDATE: ActionType = 'ENTITIES/UPDATE'

// songs
export const SONG_CURRENT: ActionType = 'SONGS/CURRENT'
export const SONG_LISTEN_REQUEST: ActionType = 'SONGS/LISTEN/REQUEST'
export const SONG_NEXT_REQUEST: ActionType = 'SONGS/NEXT/REQUEST'
export const SONG_MARK_REQUEST: ActionType = 'SONGS/MARK/REQUEST'
export const SONG_MARK_SUCCESS: ActionType = 'SONGS/MARK/SUCCESS'
export const SONG_LIKE_REQUEST: ActionType = 'SONGS/LIKE/REQUEST'
export const SONG_LIKE_SUCCESS: ActionType = 'SONGS/LIKE/SUCCESS'
export const SONG_DISLIKE_REQUEST: ActionType = 'SONGS/DISLIKE/REQUEST'
export const SONG_DISLIKE_SUCCESS: ActionType = 'SONGS/DISLIKE/SUCCESS'
