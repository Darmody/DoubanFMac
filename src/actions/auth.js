// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const login = (username: string, password: string): Action => ({
  type: types.LOGIN_REQUEST,
  payload: {
    username,
    password,
  }
})

export const logout = (): Action => ({
  type: types.LOGOUT,
  payload: undefined,
})

export const logined = (response: Object): Action => ({
  type: types.LOGIN_SUCCESS,
  payload: response,
})
