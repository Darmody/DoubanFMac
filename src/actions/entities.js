// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const save = (data: Object): Action => ({
  type: types.ENTITIES_SAVE,
  payload: data,
})

export const update = (data: Object): Action => ({
  type: types.ENTITIES_UPDATE,
  payload: data,
})
