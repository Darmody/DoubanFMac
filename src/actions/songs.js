// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const listen = (channel: number): Action => ({
  type: types.SONG_LISTEN_REQUEST,
  payload: channel,
})

export const next = (channel: number, sid?: number): Action => ({
  type: types.SONG_NEXT_REQUEST,
  payload: {
    channel,
    sid,
  }
})

export const mark = (channel: number, sid: number): Action => ({
  type: types.SONG_MARK_REQUEST,
  payload: {
    sid,
    channel,
  }
})

export const current = (song: {}): Action => ({
  type: types.SONG_CURRENT,
  payload: song,
})
