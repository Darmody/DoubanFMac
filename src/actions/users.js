// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const current = (): Action => ({
  type: types.USER_CURRENT_REQUEST,
  payload: undefined,
})
