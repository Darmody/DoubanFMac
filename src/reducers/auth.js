// @flow
import Immutable from 'seamless-immutable'
import * as types from 'constants/types/ActionTypes'
import type { Reducer } from 'constants/types/Redux'

const initialState = Immutable({
  id: null,
  token: '',
  refreshToken: '',
  expiredAt: null,
})

const reducer: Reducer = (state = initialState, { type, payload }) => {
  switch (type) {
    case types.LOGINED: return state.merge({
      token: payload.access_token,
      refreshToken: payload.refresh_token,
      id: payload.douban_user_id,
      expiredAt: payload.expires_in,
    })
    default: return state
  }
}

export default reducer
