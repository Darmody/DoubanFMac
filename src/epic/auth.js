// @flow
import { LOGIN } from 'constants/types/ActionTypes'
import type { Epic } from 'constants/types/Redux'
import { logined } from 'actions/auth'
import { fetch$ } from 'utils/observable'

const loginEpic: Epic = (action$, store, { authorize }) =>
  action$.ofType(LOGIN)
    .mergeMap(({ payload }) =>
      fetch$(authorize, payload.username, payload.password)
        .map(response => logined(response))
    )

export default [
  loginEpic
]
