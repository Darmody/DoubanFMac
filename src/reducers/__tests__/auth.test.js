import Immutable from 'seamless-immutable'
import * as types from 'constants/types/ActionTypes'
import * as actions from 'actions/auth'
import reducer from 'reducers/auth'

describe('reducer::auth', () => {
  it('initialize state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot()
  })

  it('存储认证信息', () => {
    const response = {
      accessToken: 'token',
      refreshToken: 'refresh token',
      doubanUserId: 'id',
      expiresIn: 123456,
    }
    expect(
      reducer(Immutable({ loginFailed: false }),
      actions.logined(response)),
    ).toMatchSnapshot()
  })

  it('标记登录失败', () => {
    const state = reducer(undefined, { type: types.LOGIN_FAILURE })
    expect(state).toMatchSnapshot()
    expect(state.loginFailed).toBeTruthy()
  })

  it('登出后重置 state', () => {
    const state = reducer(Immutable({ id: 1 }), { type: types.LOGOUT })
    expect(state).toMatchSnapshot()
    expect(state.id).toBeNull()
  })
})
