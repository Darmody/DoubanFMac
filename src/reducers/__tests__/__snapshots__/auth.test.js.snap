// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer::auth initialize state 1`] = `
Object {
  "expiredAt": null,
  "id": null,
  "refreshToken": "",
  "token": "",
}
`;

exports[`reducer::auth 存储认证信息 1`] = `
Object {
  "expiredAt": 123456,
  "id": "id",
  "refreshToken": "refresh token",
  "token": "token",
}
`;
